### PR TITLE
Tdd approach

### PR DESCRIPTION
**Pull request by ghost**
Hello our **Code Reviewer,**  😄
We have completed the TDD project. In this Pull Request, @baqar-abbas  and I, have done test-driven development (TDD), writing tests first and then creating the methods. 

## General requirements

- [No linter errors](https://github.com/microverseinc/linters-config). ✅ ✅ ✅
- Correct  use [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).  ✅ ✅ ✅
- Documented [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).  ✅ ✅ ✅

## Ruby requirements
- Follow [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).  ✅ ✅ ✅

### Project requirements

**1. Created first the tests of the following methods: ✅**

-  `factorial` that takes one argument, an integer `N`, and returns the [factorial](https://en.wikipedia.org/wiki/Factorial) for that number. The factorial is the multiplication of all integers from 1 to `N` and has the special case that the factorial of 0 is 1. This method only accepts 0 and positive integers, so if a negative integer is given it should raise an exception.
-  `reverse` that takes one argument, a string `word`, and returns `word` reversed (e.g. if `word` is `"hello"` it returns `"olleh"`).
- `fizzbuzz` that takes one argument, an integer `N`, and returns a string. The returned string is constructed following these rules:
  - When `N` is divisible by 3, return `"fizz"`.
  - When `N` is divisible by 5, return `"buzz"`.
  - When `N` is divisible by 3 and 5, return `"fizzbuzz"`.
  - Any other case, return `N` as a string (e.g. say `N` is `7` then return `"7"`).
 
**2. Created a class called Solver✅**
**3. Created the three methods described above inside of the class✅**

 That is all. We hope you liked it!!

***If you have any questions, don't hesitate to contact us by Slack or Zoom**

### CODE REVIEW

File: spec/factorial_spec.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+require 'rspec/mocks'
+
+describe Solver do
+  let(:solver) { Solver.new }
+
+  describe '#factorial' do
+    it 'returns zero if the parameter is zero' do
+      allow(solver).to receive(:factorial).with(0).and_return(0)
+
+      expect(solver.factorial(0)).to eq(0)
```
Comment (julie-ify) [inline spec/factorial_spec.rb]: - [ ] When the number is 0, the result should be 1, not 0. Kindly fix this.

**Review by julie-ify — CHANGES_REQUESTED** [review]
## Hi @rivasbolinga and @baqar-abbas,

Good job so far!

## Highlights!
- Good feature branch name ✔️ 
- No linter errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 
- Good code organization and working app ✔️
- Descriptive README ✔️ 

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

# Status: Required Changes ♻️⛔️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@julie-ify` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/factorial_spec.rb (hunk 1 above)
Comment (baqar-abbas) [reply spec/factorial_spec.rb]: Suggested changes by the code reviewer have been incorporated. Screen shot attached below for your kind reference.

image

Comment (baqar-abbas) [reply spec/factorial_spec.rb]: Send it for a review again

**Review by t-yanick — APPROVED** [review]
# STATUS: MILESTONE APPROVED :1st_place_medal: 

Hi @baqar-abbas and @rivasbolinga , 

### Great job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Requested changes implemented 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by baqar-abbas** [comment]
Thanks @t-yanick :)

**Pull request merged**